Add separator version of DT_FOREACH* macros

gmarull opened the pull request:
It is frequent to see in Devicetree code constructs like:

```c
#define LABEL_AND_COMMA(node_id) DT_LABEL(node_id),

const char *child_labels[] = {
	DT_FOREACH_CHILD(DT_NODELABEL(n), LABEL_AND_COMMA)
};
```

That is, an auxiliary macro to append a separator character in
DT_FOREACH* macros. Non-DT API, e.g. FOR_EACH(), takes a separator
argument to avoid such intermediate macros.

This patch adds DT_FOREACH_CHILD_SEP (and instance version
of it). It takes an extra argument: a separator. With this change, the
example above can be simplified to:

```c
const char *child_labels[] = {
	DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), DT_LABEL, (,))
};
```

Notes:
- Other DT_FOREACH* macros could/should be extended as well
- We could modify original macros to always take a separator, to align
  with non-DT FOREACH* macros, but that would be a breaking change
- Placement of `sep` argument: `DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), DT_LABEL, (,))` or `DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), (,), DT_LABEL)`? (relevant for the VARGS case)

gmarull committed:
scripts: dts: gen_defines: add DT_DEBRACKET_INTERNAL helper

Add a utility macro used to remove brackets from around a single
argument. While __DEBRACKET exists in util_internal.h, this change makes
DT independent (otherwise we should include util_internal.h, which is
another option).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
devicetree: add DT_(INST_)FOREACH_CHILD(_STATUS_OKAY)_SEP(_VARGS)

It is frequent to see in Devicetree code constructs like:

```c
 #define NAME_AND_COMMA(node_id) DT_NODE_FULL_NAME(node_id),

const char *child_names[] = {
	DT_FOREACH_CHILD(DT_NODELABEL(n), NAME_AND_COMMA)
};
```

That is, an auxiliary macro to append a separator character in
DT_FOREACH* macros. Non-DT API, e.g. FOR_EACH(), takes a separator
argument to avoid such intermediate macros.

This patch adds DT_FOREACH_CHILD_SEP (and instance/status okay/vargs
versions of it). They all take an extra argument: a separator. With this
change, the example above can be simplified to:

```c
const char *child_labels[] = {
	DT_FOREACH_CHILD(DT_NODELABEL(n), DT_NODE_FULL_NAME, (,))
};
```

Notes:
- Other DT_FOREACH* macros could/should be extended as well

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: led: pwm: use DT_INST_FOREACH_CHILD_SEP

Remove the need for auxiliary macros by using DT_INST_FOREACH_CHILD_SEP.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

galak commented:
I think its best to keep the original macro as is and add a new macro.

Do we need a version for `DT_FOREACH_CHILD_VARGS`?

gmarull commented:
> I think its best to keep the original macro as is and add a new macro.
> 
> Do we need a version for `DT_FOREACH_CHILD_VARGS`?

Yes, definitely. `DT_FOREACH_PROP_ELEM` probably as well.

gmarull commented:
> Do we have any other in-tree usages? My only concern is how often is this pattern used to introduce a new API?

Grepping for `AND_COMMA` shows a few, there are others with other namings, e.g. in `subsys/pm/state.c`.

gmarull committed:
drivers: led: gpio: use DT_INST_FOREACH_CHILD_SEP_VARGS

Avoid auxiliary macros by using DT_INST_FOREACH_CHILD_SEP_VARGS

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
arch: arm64: core: smp: use DT_FOREACH_CHILD_STATUS_OKAY_SEP

Avoid auxiliary macros by using DT_FOREACH_CHILD_STATUS_OKAY_SEP.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
devicetree: add DT_INST_FOREACH_CHILD_STATUS_OKAY* macros

Add the instance version of the DT_FOREACH_CHILD_STATUS_OKAY* family of
macros.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: gpio: nct38xx: use DT_INST_FOREACH_CHILD_STATUS_OKAY_SEP

Simplify driver by using DT_INST_FOREACH_CHILD_STATUS_OKAY_SEP (avoids
both DT_DRV_INST and auxiliary macro).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: gpio: xlnx_ps: use DT_INST_FOREACH_CHILD_STATUS_OKAY_SEP

Simplify driver by using DT_INST_FOREACH_CHILD_STATUS_OKAY_SEP (avoids
DT_DRV_INST and auxiliary macro).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: intc: gic: use DT_FOREACH_CHILD_STATUS_OKAY_SEP

Remove the need for an auxiliary macro by using
DT_FOREACH_CHILD_STATUS_OKAY_SEP.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
samples: drivers: led_pwm: use DT_FOREACH_CHILD_SEP_VARGS

use DT_FOREACH_CHILD_SEP_VARGS to avoid an auxiliary macro.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: state: use DT_FOREACH_CHILD_SEP

Avoid usage of auxiliary macros by using DT_FOREACH_CHILD_SEP.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
> You asked about argument order offline -- I don't see any specific questions on that in the thread here but it looks generally OK to me.

It's in the PR description:

> Placement of sep argument: DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), DT_LABEL, (,)) or DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), (,), DT_LABEL)? (relevant for the VARGS case)

The reason is that with current approach, separator will go between function and VARGS, which may be confusing. If it goes before, we'll have separator + function + vargs.

mbolivar-nordic commented:
I personally think this order looks good:

```
DT_FOREACH_FOO_SEP(node_id, fn, sep)
DT_FOREACH_FOO_SEP_VARGS(node_id, fn, sep, ...)
```

I'm not married to it, though

mbolivar-nordic commented on include/zephyr/devicetree.h:
```diff
@@ -1927,6 +1927,43 @@
+ * Example usage:
+ *
+ *     const char *child_labels[] = {
+ *         DT_FOREACH_CHILD_SEP(DT_NODELABEL(n), DT_LABEL, (,))
```
Let's not use DT_LABEL given the removal of those properties. How about DT_NODE_FULL_NAME instead?

gmarull replied:
updated

mbolivar-nordic commented on subsys/pm/state.c:
```diff
@@ -39,23 +39,21 @@ BUILD_ASSERT(DT_NODE_EXISTS(DT_PATH(cpus)),
 /* Check that all power states are consistent */
 DT_FOREACH_CHILD(DT_PATH(cpus), CHECK_POWER_STATES_CONSISTENCY)
 
-#define NUM_CPU_STATES(n) DT_NUM_CPU_POWER_STATES(n),
```
Looks like DT_NUM_CPU_POWER_STATES was added by you in 7bfd0976aad, with a potential move to devicetree.h in the future. Gentle reminder that this is still an abuse of the DT_ namespace.

gmarull replied:
thanks for the reminder here, will likely move that macro to DT API

gmarull commented on scripts/dts/gen_defines.py:
```diff
@@ -267,6 +269,13 @@ def write_top_comment(edt):
+    # Writes utility macros
+
+    out_comment("Used to remove brackets from around a single argument")
+    out_define("__DT_DEBRACKET(...)", "__VA_ARGS__")
```
done
